rewatch/src/project_context.rs

```diff
@@ -0,0 +1,288 @@
+use crate::build::packages;
+use crate::config::Config;
+use crate::helpers;
+use ahash::AHashSet;
+use anyhow::Result;
+use anyhow::anyhow;
+use log::debug;
+use std::fmt;
+use std::path::{Path, PathBuf};
+
+type RescriptJsonPath = PathBuf;
+
+/// Represents the context of a command, categorizing in various ways:
+/// - SingleProject: one rescript.json with no parent or children.
+/// - MonorepoRoot: a rescript.json with (dev-)dependencies found in subfolders.
+/// - MonorepoPackage: a rescript.json with a parent rescript.json.
+///   The current package name is listed in the parent (dev-)dependencies.
+// I don't think this linting rule matters very much as we only create a single instance of this enum.
+// See https://rust-lang.github.io/rust-clippy/master/index.html#large_enum_variant
+#[allow(clippy::large_enum_variant)]
+pub enum ProjectContext {
```

Comment:
This streamlines the many `workspace: Option<string>` checks that were being done.

It improves the readability and debugability of rewatch.

Reply:
I would maybe do:
```
pub enum MonoRepoContext {
  Root {local_dependencies, local_dev_dependencies}
  Package {parent_config: Config, parent_path: RescriptJsonPath}
}

pub struct ProjectContext {
  config: Config,
  path: RescriptJsonPath,
  mono_repo_context: Option<MonoRepoContext>
}
```

I would like to be mindful about memory consumption - even if minor - you never know how much these will grow in the future - and I think we should try and keep a small footprint (smaller than we do now). I think the `enum` will otherwise always take the biggest size when allocating, even in smaller contexts.